src/Input/AbstractParamAwareInput.php

```diff
@@ -0,0 +1,203 @@
+    {
+        $this->input  = $input;
+        $this->output = $output;
+        $this->helper = $helper;
+        $this->params = $params;
+    }
+
+    /**
+     * Define this method in order to modify the question, if needed, before
+     * prompting for an answer.
+     */
+    abstract protected function modifyQuestion(Question $question): void;
+
+    /**
+     * @return mixed
+     * @throws InvalidArgumentException When the parameter does not exist.
+     * @throws InvalidArgumentException When the parameter is of an invalid type.
+     * @throws InvalidArgumentException When the parameter is required, input is
+     *     non-interactive, and no value is provided.
+     */
+    public function getParam(string $name)
+    {
+        if (! is_array($this->params) || ! isset($this->params[$name])) {
+            throw new InvalidArgumentException(sprintf('Invalid parameter name: %s', $name));
+        }
+
+        $value      = $this->input->getOption($name);
+        $inputParam = $this->params[$name];
+
+        if (! $inputParam instanceof InputParamInterface) {
+            throw new InvalidArgumentException(sprintf(
+                'Invalid parameter type; must be of type %s',
+                InputParamInterface::class
+            ));
+        }
+
+        $question = $inputParam->getQuestion();
+
+        $this->modifyQuestion($question);
+
+        if ($value === null && ! $this->input->isInteractive()) {
+            $value = $inputParam->getDefault();
+        }
+
+        if ($value !== null) {
+            $validator = $question->getValidator();
+            if ($validator) {
+                $validator($value);
+            }
+
+            $normalizer = $question->getNormalizer();
+
+            return $normalizer === null ? $value : $normalizer($value);
+        }
+
+        if (! $this->input->isInteractive() && $inputParam->isRequired()) {
+            throw new InvalidArgumentException(sprintf('Missing required value for --%s parameter', $name));
+        }
+
+        $value = $this->helper->ask($this, $this->output, $question);
+
+        // set the option value so it can be reused in chains
+        $this->input->setOption($name, $value);
+
+        return $value;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return null|string
+     */
+    public function getFirstArgument()
+    {
+        return $this->input->getFirstArgument();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function bind(InputDefinition $definition)
+    {
+        $this->input->bind($definition);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function validate()
+    {
+        $this->input->validate();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return array
+     */
+    public function getArguments()
+    {
+        return $this->input->getArguments();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @param string|int $name
```

Comment:
Honestly, I did this because intelephense was not able to provide me with type hints otherwise. I'm not sure if that's a problem specific to intelephense, though.

Reply:
Np, its not wrong and until we are using phpstan, this might not be a problem at all.
But when adding phpstan, we will get issues due to the reason I've stated above.